### PR TITLE
Ingela/ssl/select appropriate cert/gh 4143/otp 15993

### PR DESCRIPTION
**Pull request by IngelaAndin**
Addresses the feature request form ERL-1016 and #4143
@essen I tried to have this prioritized for quite a while now, but better late than never. Hope it will be what you wished for.

### CODE REVIEW

File: lib/ssl/doc/src/ssl.xml [hunk 1]
```diff
@@ -356,13 +356,53 @@
...
+	</p>
+
+	<note><p> <c>eddsa</c> certificates are only supported by TLS-1.3 that does not support  <c>dsa</c> certificates.
+	 <c>rsa_pss_pss</c> (RSA certificates using Probabilistic Signature Scheme) are supported in TLS-1.2 and TLS-1.3, but some
+	TLS-1.2 implementations may not support <c>rsa_pss_pss</c>.
+	</p></note>
+      </desc>
+    </datatype>
+
+    <datatype>
```
Comment (KennethL) [inline lib/ssl/doc/src/ssl.xml:397]: There must be examples for how this is to be used and also how this option can replace the use of the old options for certificates and keys

Comment (KennethL) [inline lib/ssl/doc/src/ssl.xml:397]: There must be examples for how this is to be used and also how this option can replace the use of the old options for certificates and keys

File: lib/ssl/src/dtls_connection.erl [hunk 2]
```diff
@@ -208,13 +208,14 @@ initial_hello({call, From}, {start, Timeout},
                                      session_cache_cb = CacheCb},
             protocol_specific = PS,
             handshake_env = #handshake_env{renegotiation = {Renegotiation, _}},
-            connection_env = #connection_env{cert_key_pairs = CertKeyPairs} = CEnv,
+            connection_env = #connection_env{cert_key_alts = CertKeyAlts} = CEnv,
```
Comment (KennethL) [inline lib/ssl/src/dtls_connection.erl:211]: I don't think this is a better name than before

Comment (KennethL) [inline lib/ssl/src/dtls_connection.erl:211]: I don't think this is a better name than before

File: lib/ssl/src/ssl.erl [hunk 3]
```diff
@@ -2480,7 +2494,7 @@ handle_hashsigns_option(_, _Version) ->
     undefined.
 
 handle_signature_algorithms_option(Value, Version) when is_list(Value)
-                                                        andalso Version >= {3, 4} ->
+                                                        andalso Version >= {3, 3} ->
```
Comment (KennethL) [inline lib/ssl/src/ssl.erl:2497]:
```suggestion
                                                        , Version >= {3, 3} ->
```
```suggestion
                                                        andalso Version >= {3, 3} ->
```
avoid `andalso` in guards when it is not necessary.

Comment (KennethL) [inline lib/ssl/src/ssl.erl:2497]:
```suggestion
                                                        , Version >= {3, 3} ->
```
```suggestion
                                                        andalso Version >= {3, 3} ->
```
avoid `andalso` in guards when it is not necessary.

File: lib/ssl/src/ssl_certificate.erl [hunk 4]
```diff
@@ -326,6 +328,56 @@ handle_cert_auths([_ | Certs] = EChain, CertAuths, _, _) ->
             {error, EChain, not_in_auth_domain}
     end.
 
+available_cert_key_pairs(CertKeyGroups) ->
+    RevAlgos = [dsa, rsa, rsa_pss_pss, ecdsa, eddsa],
+    cert_key_group_to_list(RevAlgos, CertKeyGroups, []).
+
+available_cert_key_pairs(CertKeyGroups, Version) ->
+    RevAlgos = [dsa, rsa, rsa_pss_pss, ecdsa, eddsa],
+    cert_key_group_to_list(RevAlgos, CertKeyGroups, Version, []).
```
Comment (KennethL) [inline lib/ssl/src/ssl_certificate.erl:337]: Why is two functions one with version and one without needed?

Comment (KennethL) [inline lib/ssl/src/ssl_certificate.erl:337]: Why is two functions one with version and one without needed?

File: lib/ssl/src/ssl_certificate.erl [hunk 5]
```diff
@@ -326,6 +328,56 @@ handle_cert_auths([_ | Certs] = EChain, CertAuths, _, _) ->
...
+
+cert_key_group_to_list([], _, Acc) ->
+    final_group_list(Acc);
+cert_key_group_to_list([Algo| Rest], CertKeyGroups, Acc) ->
```
Comment (KennethL) [inline lib/ssl/src/ssl_certificate.erl:341]: Need an explanation if this whole function and what it is supposed to do.

Comment (KennethL) [inline lib/ssl/src/ssl_certificate.erl:341]: Need an explanation if this whole function and what it is supposed to do.

File: lib/ssl/src/ssl_handshake.erl [hunk 6]
```diff
@@ -1047,24 +1048,25 @@ prf({3,_N}, PRFAlgo, Secret, Label, Seed, WantedLength) ->
     {ok, tls_v1:prf(PRFAlgo, Secret, Label, Seed, WantedLength)}.
 
 select_session(SuggestedSessionId, CipherSuites, HashSigns, Compressions, SessIdTracker, Session0, Version, SslOpts, CertKeyPairs) ->
+    CertKeyAlts = ssl_certificate:available_cert_key_pairs(CertKeyPairs, Version),
```
Comment (KennethL) [inline lib/ssl/src/ssl_handshake.erl:1051]: Need a walkthrough of this code

Comment (KennethL) [inline lib/ssl/src/ssl_handshake.erl:1051]: Need a walkthrough of this code

File: lib/ssl/src/ssl.erl [hunk 7]
```diff
@@ -349,8 +350,14 @@
                                        key_id := crypto:key_id(), 
                                        password => crypto:password()}. % exported
 -type key_pem()                   :: file:filename().
--type key_password()              :: iodata() | fun(() -> iodata()).
--type cipher_suites()             :: ciphers().    
+-type key_pem_password()          :: iodata() | fun(() -> iodata()).
+-type certs_keys()                :: [cert_key_pair()].
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/ssl.erl:354]: Well, naming is hard, that was actually my first suggestion. But there were some other concerns about pushing the name pair too much as there is also other information included in addition to the actual certificate and its associated key (Possible chain certs and password for the key). This name tries to reflect that it is a set of certificates and keys. We can think about it some more.

File: lib/ssl/src/ssl_certificate.erl (hunk 5 above)
Comment (IngelaAndin) [reply lib/ssl/src/ssl_certificate.erl:341]: The inputted list of cert and key pairs is in arbitrary order and may represent all or some of the possible signing algorithms. Internally this list will be grouped in different lists (one per signing algorithm where each list can be sorted on the relative key strength unless it is an engine key). Then later we will use these groups the create one prioritized list that we can use the first one supported by negotiate algorithms and other relevant criteria.

File: lib/ssl/src/ssl.erl [hunk 8]
```diff
@@ -349,8 +350,14 @@
...
+-type cert_key_pair()             :: #{cert => cert(),
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/ssl.erl:355]: That feels backward, why should the type make the user supply different keys for the conceptual same thing?  I agree that the old options are sort of doing that as they are inherited from OpenSSL.

File: lib/ssl/doc/src/ssl.xml (hunk 1 above)
Comment (IngelaAndin) [reply lib/ssl/doc/src/ssl.xml:397]: Will add something in the user guide.

File: lib/ssl/src/ssl.erl (hunk 8 above)
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/ssl.erl:355]: This has been discussed some more and will be revisited.

[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/ssl.erl:355]: I think we both potentially misunderstood each other, please check added examples.

File: lib/ssl/doc/src/ssl.xml [hunk 9]
```diff
@@ -356,13 +356,53 @@
...
+      <name name="certs_keys"/>
+      <desc>
+	<p>A list of a certificate (or possible a certificate and its
+	chain) and the associated key of the certificate, that may be
+	used to authenticate the client or the server.  The
+	certificate key pair that is considered best and matches
+	negotiated parameters for the connection will be selected.
+	Different signature algorithms are prioritized in the order
+	<c> eedsa, ecdsa, rsa_pss_pss, rsa and dsa </c>.  If more than
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/doc/src/ssl.xml:375]: fixed

File: lib/ssl/doc/src/ssl.xml [hunk 10]
```diff
@@ -356,13 +356,53 @@
...
+	one key is supplied for the same signing algorithm (which is
+	probably an unusual use case) they will prioritized by
+	strength unless it is a so called <c>engine key</c> that will
+	be favoured over other keys. As engine keys can not be
+	inspected suppling more than one engine key will make no
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/doc/src/ssl.xml:380]: fixed

File: lib/ssl/doc/src/ssl.xml [hunk 11]
```diff
@@ -356,13 +356,53 @@
...
+	sense. This offers flexibility to for instance configure a
+	newer certificate that is expected to be used in most cases
+	and an older but acceptable certificate that will only be used
+	to communicate with legacy systems. Note that there is a trade
+	off between the induced overhead and the flexibility so
+	alternatives should be chosen for good reasons. If specified
+	overrides all single certificate and key options. For examples see <seeguide marker="ssl:using_ssl"> the Users Guide</seeguide>
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/doc/src/ssl.xml:387]: Changed it to 
If the <c>certs_keys</c> option is specified it overrides all single certificate and key options.

File: lib/ssl/doc/src/using_ssl.xml [hunk 12]
```diff
@@ -462,6 +469,23 @@ ok
       {ok, CSock} = ssl:transport_accept(LSock).
     </code>
 
+
+    <p><em>Step 1 (server):</em> with alternative certificates, in this example the EDDSA certificate will be prefered if TLS-1.3 is negotiated and the RSA certificate will always be used for TLS-1.2 as it does not support the EEDSA algorithm: Added in OTP-25</p>
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/doc/src/using_ssl.xml]: Hum ... looks better rendered now but the XML looks worse :-/

[the remark this answers is not in the document]
Comment (u3s) [reply lib/ssl/doc/src/using_ssl.xml]: Leave it, if a problem.

File: lib/ssl/test/ssl_api_SUITE.erl [hunk 13]
```diff
@@ -3169,3 +3191,183 @@ dtls_exclusive_non_default_version(DTLSVersion) ->
...
+                          {verify, verify_peer}, {versions, ['tlsv1.2']},
+                          {cacerts, SEDDSACACerts ++ SECDSACACerts}],
+        client_config => [{certs_keys, ClientCertKeys},
+                          {verify, verify_peer},  {versions, ['tlsv1.2']},
+                          {cacerts, CEDDSACACerts ++ CECDSACACerts}]},
+      {client_peer, SECDSACert}, {server_peer, CECDSACert}}
+    ];
+test_config('tlsv1.2', _) ->
+    #{server_config := SRSAOpts,
+      client_config := CRSAOpts} = public_key:pkix_test_data(#{server_chain => #{root => ssl_test_lib:eddsa_conf(),
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/test/ssl_api_SUITE.erl:3263]: It is proper, that is we want to check that rsa certs will be picked as eddsa is not supported in TLS-1.2

File: lib/ssl/test/ssl_api_SUITE.erl [hunk 14]
```diff
@@ -3169,3 +3191,183 @@ dtls_exclusive_non_default_version(DTLSVersion) ->
...
+                                         {mfa, {?MODULE, check_peercert, [ExpectedClient]}},
+                                         {options,  ssl_test_lib:ssl_options(ClientOpts, Config)}
+					]),
+
+    ssl_test_lib:check_result(Server, ok, Client, ok),
+    %% Make sure to start next test fresh
+    ssl:stop(),
+    ssl:start().
+
+test_config('tlsv1.3', _) ->
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/test/ssl_api_SUITE.erl:3217]: Well, test code easily becomes a little cut and pasty I think. I will try your wrapper idea to see if feels better.

File: lib/ssl/doc/src/using_ssl.xml (hunk 12 above)
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/doc/src/using_ssl.xml]: Hade to remove some of it was not allowed by xmlint!

**Comment by IngelaAndin** [comment]
Merging this for OTP-25 release candidate 3. We will add more tests and documentation for the final release. Maybe also minor updates to naming conventions.

**Comment by essen** [comment]
Hello! Catching up on OTP PRs, just noticed this, sorry I missed it! We will look it up in details soon. Thanks for the hard work!

**Comment by IngelaAndin** [comment]
Thanks, heads up I am working on some adjustments at the moment for the release. But the concept will not change.

**Comment by IngelaAndin** [comment]
Adjustments where made in 165a3368dafcf3b8290f2658f27ffa885c145b9.
It probably will be some more doc updates that will come for the release. But hopefully, the code is good now.

**Pull request merged**